### PR TITLE
docs: make pubsub howto consistent with the new howto template

### PR DESCRIPTION
**Pull request by eliben**
Had to reorder things a big because of the way this howto was originally structured. If this looks good, I'll update the subscribe page similarly

### CODE REVIEW

File: internal/website/content/howto/pubsub/publish.md [hunk 1]
```diff
@@ -7,40 +7,40 @@ toc: true
...
-## Constructors versus URL openers
+## Opening a Topic {#opening}
 
 The easiest way to open a topic is using [`pubsub.OpenTopic`][] and a URL
 pointing to the topic, making sure you ["blank import"][] the driver package to
 link it in. See [Concepts: URLs][] for more details. If you need fine-grained
 control over the connection settings, you can call the constructor function in
-the driver package directly (like `gcppubsub.OpenTopic`). This guide will show
-how to use both forms for each pub/sub provider.
+the driver package directly (like `gcppubsub.OpenTopic`). [This guide][] will
```
Comment (vangent) [inline internal/website/content/howto/pubsub/publish.md]: I think the blob howto worded this better:

See the [guide below][] for usage of both forms for each supported provider.

File: internal/website/content/howto/pubsub/publish.md [hunk 2]
```diff
@@ -7,40 +7,40 @@ toc: true
...
+[This guide]: {{< ref "#services" >}}
 [`pubsub.OpenTopic`]:
 https://godoc.org/gocloud.dev/pubsub#OpenTopic
 ["blank import"]: https://golang.org/doc/effective_go.html#blank_import
 [Concepts: URLs]: {{< ref "/concepts/urls.md" >}}
 
-## Amazon Simple Notification Service {#sns}
+## Sending messages on a Topic {#sending}
+
+Sending a message on a Topic looks like this:
```
Comment (vangent) [inline internal/website/content/howto/pubsub/publish.md]: Add a link to `*pubsub.Topic` godoc here?

Topic -> [*pubsub.Topic][]

[*pubsub.Topic]: https://godoc.org/gocloud.dev/pubsub#Topic

File: internal/website/content/howto/pubsub/publish.md (hunk 1 above)
Comment (eliben) [reply internal/website/content/howto/pubsub/publish.md]: Done

File: internal/website/content/howto/pubsub/publish.md (hunk 2 above)
Comment (eliben) [reply internal/website/content/howto/pubsub/publish.md]: Done

File: internal/website/content/howto/pubsub/publish.md [hunk 3]
```diff
@@ -7,40 +7,40 @@ toc: true
...
+
+{{< goexample src="gocloud.dev/pubsub.ExampleTopic_Send" imports="0" >}}
+
+Note that the [semantics of message delivery][] can vary by provider.
+
+[semantics of message delivery]: https://godoc.org/gocloud.dev/pubsub#hdr-At_most_once_and_At_least_once_Delivery
+
+## Supported Pub/Sub Services {#services}
+
+### AWS Simple Notification Service {#sns}
```
Comment (polinasok) [inline internal/website/content/howto/pubsub/publish.md]: According to Google the long name with all words spelled out starts with Amazon, not AWS. The abbreviated version is used with as both Amazon SNS and AWS SNS.

File: internal/website/content/howto/pubsub/publish.md [hunk 4]
```diff
@@ -73,7 +73,7 @@ create an [AWS session][] with the same region as your topic:
 [`awssnssqs.OpenSNSTopic`]: https://godoc.org/gocloud.dev/pubsub/awssnssqs#OpenSNSTopic
 [AWS session]: https://docs.aws.amazon.com/sdk-for-go/api/aws/session/
 
-## Amazon Simple Notification Service {#sqs}
+### AWS Simple Queue Service {#sqs}
```
Comment (polinasok) [inline internal/website/content/howto/pubsub/publish.md]: Same suggestion as above

File: internal/website/content/howto/pubsub/publish.md [hunk 5]
```diff
@@ -106,7 +106,7 @@ create an [AWS session][] with the same region as your topic:
 [`awssnssqs.OpenSQSTopic`]: https://godoc.org/gocloud.dev/pubsub/awssnssqs#OpenSQSTopic
 [AWS session]: https://docs.aws.amazon.com/sdk-for-go/api/aws/session/
 
-## Google Cloud Pub/Sub {#gcp}
+### GCP Cloud Pub/Sub {#gcp}
```
Comment (polinasok) [inline internal/website/content/howto/pubsub/publish.md]: Isn't this called Google Cloud Pub/Sub? Why the change?

File: internal/website/content/howto/pubsub/publish.md [hunk 6]
```diff
@@ -7,40 +7,40 @@ toc: true
...
+show how to use both forms for each pub/sub provider.
 
+[This guide]: {{< ref "#services" >}}
 [`pubsub.OpenTopic`]:
 https://godoc.org/gocloud.dev/pubsub#OpenTopic
 ["blank import"]: https://golang.org/doc/effective_go.html#blank_import
 [Concepts: URLs]: {{< ref "/concepts/urls.md" >}}
 
-## Amazon Simple Notification Service {#sns}
+## Sending messages on a Topic {#sending}
```
Comment (polinasok) [inline internal/website/content/howto/pubsub/publish.md]: Please capitalize "Messages"

File: internal/website/content/howto/pubsub/publish.md [hunk 7]
```diff
@@ -7,40 +7,40 @@ toc: true
 
 Publishing a message to a topic with the Go CDK takes two steps:
 
-1. Open the topic with the Pub/Sub provider of your choice (once per topic).
-2. Send messages on the topic.
+1. [Open the topic][] with the Pub/Sub provider of your choice (once per topic).
+2. [Send messages][] on the topic.
```
[the remark this answers is not in the document]
Comment (eliben) [reply internal/website/content/howto/pubsub/publish.md]: This is already state on the landing page for pubsub (_index.md): https://gocloud.dev/howto/pubsub/

I still like the bullets here since they provide a brief intro to the rest of the page. I don't think the TOC is a replacement

File: internal/website/content/howto/pubsub/publish.md [hunk 8]
```diff
@@ -7,40 +7,40 @@ toc: true
...
 
-<!--more-->
```
[the remark this answers is not in the document]
Comment (eliben) [reply internal/website/content/howto/pubsub/publish.md]: Oops, nice catch.

File: internal/website/content/howto/pubsub/publish.md (hunk 6 above)
Comment (eliben) [reply internal/website/content/howto/pubsub/publish.md]: Done

File: internal/website/content/howto/pubsub/publish.md [hunk 9]
```diff
@@ -7,40 +7,40 @@ toc: true
...
 [Concepts: URLs]: {{< ref "/concepts/urls.md" >}}
 
-## Amazon Simple Notification Service {#sns}
+## Sending messages on a Topic {#sending}
+
+Sending a message on a Topic looks like this:
+
+{{< goexample src="gocloud.dev/pubsub.ExampleTopic_Send" imports="0" >}}
+
+Note that the [semantics of message delivery][] can vary by provider.
```
[the remark this answers is not in the document]
Comment (eliben) [reply internal/website/content/howto/pubsub/publish.md]: Done

File: internal/website/content/howto/pubsub/publish.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (eliben) [reply internal/website/content/howto/pubsub/publish.md]: Done

Comment (eliben) [reply internal/website/content/howto/pubsub/publish.md]: Will change to Amazon here

File: internal/website/content/howto/pubsub/publish.md [hunk 10]
```diff
@@ -63,7 +63,7 @@ you may need to manually Base64 decode the message payload.
 [SQS Subscribe]: {{< relref "./subscribe.md#sqs" >}}
 [SNS]: https://aws.amazon.com/sns/
 
-### Amazon Simple Notification Service Constructor {#sns-ctor}
+#### Amazon Simple Notification Service Constructor {#sns-ctor}
```
[the remark this answers is not in the document]
Comment (eliben) [reply internal/website/content/howto/pubsub/publish.md]: OK, so SNS just here for the constructor, but fully spelled out in the parent header?

File: internal/website/content/howto/pubsub/publish.md (hunk 4 above)
Comment (eliben) [reply internal/website/content/howto/pubsub/publish.md]: Done

File: internal/website/content/howto/pubsub/publish.md (hunk 5 above)
Comment (eliben) [reply internal/website/content/howto/pubsub/publish.md]: Changing back

**Comment by eliben** [comment]
Thanks for the careful review, @polinasok PTAL

File: internal/website/content/howto/pubsub/publish.md [hunk 11]
```diff
@@ -13,34 +13,59 @@ Publishing a message to a topic with the Go CDK takes two steps:
...
+The Go CDK can publish to a Google [Cloud Pub/Sub][] topic. The URLs use the
+project ID and the topic ID. `pubsub.OpenTopic` will use [Application Default
+Credentials][GCP creds].
+
+{{< goexample "gocloud.dev/pubsub/gcppubsub.Example_openTopicFromURL" >}}
+
+[Cloud Pub/Sub]: https://cloud.google.com/pubsub/docs/
+[GCP creds]: https://cloud.google.com/docs/authentication/production
+
+#### GCP Cloud Pub/Sub Constructor {#gcp-ctor}
```
Comment (polinasok) [inline internal/website/content/howto/pubsub/publish.md]: I think you forgot to change this GCP back.

Comment (eliben) [reply internal/website/content/howto/pubsub/publish.md]: Done

**Comment by eliben** [comment]
I've updated the "subscribe" part to be aligned with the publish now

File: internal/website/content/howto/pubsub/subscribe.md [hunk 12]
```diff
@@ -7,48 +7,73 @@ toc: true
 
 Subscribing to messages on a topic with the Go CDK takes three steps:
 
-1. Open the subscription with the Pub/Sub provider of your choice (once per subscription).
+1. Open the subscription with the Pub/Sub provider of your choice (once per
```
Comment (polinasok) [inline internal/website/content/howto/pubsub/subscribe.md]: Make these links to the subsequent sections like in the other file?

Comment (eliben) [reply internal/website/content/howto/pubsub/subscribe.md]: Done, and changed to 2 steps as discussed

File: internal/website/content/howto/pubsub/subscribe.md [hunk 13]
```diff
@@ -7,12 +7,14 @@ toc: true
 
 Subscribing to messages on a topic with the Go CDK takes three steps:
 
-1. Open the subscription with the Pub/Sub provider of your choice (once per
+1. [Open the subscription][] with the Pub/Sub provider of your choice (once per
```
Comment (polinasok) [inline internal/website/content/howto/pubsub/subscribe.md]: I think "Open a subscription" is more grammatically correct

File: internal/website/content/howto/pubsub/subscribe.md [hunk 14]
```diff
@@ -7,12 +7,14 @@ toc: true
...
    subscription).
-2. Receive messages from the topic.
-3. For each message, acknowledge its receipt using the `Ack` method
-   after completing any work related to the message. This will prevent the
-   message from being redelivered.
+2. [Receive messages][] from the topic. For each message, acknowledge its receipt
```
Comment (polinasok) [inline internal/website/content/howto/pubsub/subscribe.md]: How about "[Receive and acknowledge messages][] from the topic. After completing any work related to the message, use the `Ack` method to prevent it from being redelivered.

File: internal/website/content/howto/pubsub/subscribe.md (hunk 13 above)
Comment (eliben) [reply internal/website/content/howto/pubsub/subscribe.md]: Just took it from older version :)

 Sure, Done.

File: internal/website/content/howto/pubsub/subscribe.md (hunk 14 above)
Comment (eliben) [reply internal/website/content/howto/pubsub/subscribe.md]: Done
